### ISSUE TITLE
Passing tags and attributes to nh3.clean()

### ISSUE THREAD

**Issue by carltongibson** [matthiask/django-prose-editor#9]
Hi @matthiask — thanks for the great package, as always 🎁

I have a question about the nh3 sanitizer: 

https://github.com/matthiask/django-prose-editor/blob/1a18e77c1ceb52f0b678e18d1c8dec8c21d267f7/django_prose_editor/sanitized.py#L4-L7

Here we’re using `clean` but I wonder if we could pass the `tags` and `attributes` arguments as well?

Deriving those from the `types` used to instantiate the prose editor would allow precisely limiting the allowed HTML, I’m hoping. 

What do you think?

**Comment by matthiask** [comment]
Thank you :)

So, I'm not 100% sure if this would be better achieved through settings, field arguments or some other way or if maybe it shouldn't be done at all.

The argument for not making the nh3 invocation configurable is that nh3's job here isn't to clean up the HTML, it's only to protect against XSS when someone circumvents the ProseMirror widget. As long as editors do not do anything strange ProseMirror will always serialize a cleaned up version of the HTML fragment already, even if you use the raw HTML popup. (That's also the reason why I use nh3 instead of something more opinionated such as html-sanitizer.)

**Comment by carltongibson** [comment]
So @adamchainz wrote a post on `nh3` in which he says: 

> The nh3 defaults (from Ammonia) are generally safe, allowing only general content tags and attributes, but they are still pretty wide, allowing 75 different tags. Allowing some tags on this default list, such as <article>, may lead to surprising results. In the worst case, an attacker may be able to craft official-looking content with evil instructions to users.
> 
> Typically, incoming HTML fragments come from a rich text editor on your site. In this case, use the arguments of nh3.clean() to limit allowed tags and attributes to only those your editor will create. This way, there won’t be any chance of surprise content.
> <small>([src](https://adamj.eu/tech/2023/12/13/django-sanitize-incoming-html-nh3/#customize-cleaning-with-nh3-clean-arguments))</small>

I find that kind of compelling, and we do know the restricted set of tags and attributes that should be allowed (I think 🤔) 

`SanitizedProseEditorField` knows the config that it will pass to the widget at `__init__()` so could determine the tags and attributes allowed and pass those parameters in a closure to the `sanitize` function (or similar). (Better than a setting no?) 

I think it's _safe_ as is, but could maybe be more so (hardened) by restricting the allowed tags. 

No sure either. But just working it through with you.

**Comment by carltongibson** [comment]
I guess a validator here might show if someone was trying to be naughty 😅

**Comment by matthiask** [comment]
> I find that kind of compelling, and we do know the restricted set of tags and attributes that should be allowed (I think 🤔)

Yes, by looking at e.g. prosemirror-schema-basic, for example the link mark:
https://github.com/ProseMirror/prosemirror-schema-basic/blob/4e358f5443f15abe09406c9327e35aa9e3c1bf8c/src/schema-basic.ts#L113-L123

I think the nodes and marks definitions in prosemirror-schema-basic and prosemirror-schema-list have been unchanged (practically) since the beginning. I have had discussions in private about allowing `id` and `name` attributes, and maybe `target` could be useful, but that's a different issue.

> SanitizedProseEditorField knows the config that it will pass to the widget at __init__() so could determine the tags and attributes allowed and pass those parameters in a closure to the sanitize function (or similar). (Better than a setting no?)

Deriving the list of allowed attributes and tags from the `types` list sounds like the way to go, I agree with that.

> I think it's safe as is, but could maybe be more so (hardened) by restricting the allowed tags.
> [...]
> I guess a validator here might show if someone was trying to be naughty 😅

I totally agree with that, and I think there's value in hardening! A little more opinions on the server side would certainly be useful. Just to spell it out for everyone else, the cleaning is already more opinionated than nh3 is since e.g. the empty pararaph `<p></p>` is replaced by an empty string so that "no content" is actually falsy instead of truthy.

**Comment by carltongibson** [comment]
Great. Makes sense. I will try and prototype something when I get a small window.

**Comment by carltongibson** [comment]
902ee3c6206015aa5e177fcf1e4773a2c27e8fe3 is relevant here. 

(@matthiask I still keep thinking about this but I'm a little way off getting to it. Don't mind if you want to close — we can always reopen if/when I get to it.)

**Comment by matthiask** [comment]
@carltongibson Me too, I'm good with keeping it open. It's a good reminder.

By the way, I just reintroduced the `config={"types": [...]}` based configuration since it could be useful for simple cases. Not everyone wants to write their own JavaScript modules. Presets make solving this issue automatically hard, but basing a sanitization allowlist on the list of types is probably doable.

**Comment by matthiask** [comment]
@carltongibson 
I think I have a solution for this. For example:

```python
class ConfigurableProseEditorModel(models.Model):
    description = ConfigurableProseEditorField(
        extensions={
            "Bold": True,
            "Italic": True,
            "Heading": {"levels": [1, 2, 3]},  # Limit to h1, h2, h3
        },
        # sanitize=True,  # this is the default
    )

    def __str__(self):
        return self.description
```

This automatically generates a really restrictive nh3 configuraton allowing only `{"p", "strong", "em", "h1", "h2", "h3}` and, in this case, no attributes at all.

The pull request (#31) is a large mess. Besides the unit tests it has only been tested very superficially, but the way how it enables us to customize the sanitization configuration and configure Tiptap extensions in Python code feels right (tm) to me.

Also, the escape hatch to do everything by hand still exists, but it's already much much more flexible than what the initial `SanitizedProseEditorField` offered while also being more secure/restrictive.

I'm not asking for a review, that wouldn't be fair. If you do find the time for a quick documentation flyover that would be awesome, but really no pressure. Reporting the issue and keeping it front and center already was a big help, so thanks for that.

**Comment by carltongibson** [comment]
Great @matthiask — I think this is a really nice refinement. 🎉
